### PR TITLE
8284880: Re-examine sun.invoke.util.Wrapper hash tables

### PR DESCRIPTION
**Pull request by cl4es**
This patch examines and optimizes `Wrapper` lookups.

First wrote a few simple microbenchmarks to verify there are actual speedups from using perfect hash tables in `sun.invoke.util.Wrapper` compared to simpler lookup mechanisms (such as if-else or switch). Turns out there _is_ a speed-up for the case of `char` -> `Wrapper`, but not when mapping from `Class<?>` -> `Wrapper`, so let's drop those. The `forPrimitiveType` case didn't use the `FROM_CHAR` table for some reason, which is remedied.  

Micros show benefits across the board for warmed up case:

```
Baseline, OOTB
Benchmark                  Mode  Cnt   Score   Error  Units
Wrappers.forBasicType      avgt    5  14.387 ? 0.127  ns/op
Wrappers.forPrimitive      avgt    5  38.818 ? 0.592  ns/op
Wrappers.forPrimitiveType  avgt    5  26.085 ? 2.291  ns/op
Wrappers.forWrapper        avgt    5  44.459 ? 1.635  ns/op

Patch, OOTB
Benchmark                  Mode  Cnt   Score   Error  Units
Wrappers.forBasicType      avgt    5  14.357 ? 0.133  ns/op
Wrappers.forPrimitive      avgt    5  23.930 ? 0.071  ns/op
Wrappers.forPrimitiveType  avgt    5  14.343 ? 0.017  ns/op
Wrappers.forWrapper        avgt    5  27.622 ? 0.022  ns/op
```

For `-Xint` case (`Wrapper` use is prominent during warmup, e.g., when spinning up of MHs) there are decent or even great wins in all cases but `forPrimitiveType` - which was changed from a simple switch to use the hash lookup. Since the interpreter penalty is small in absolute terms and the win on JITed code is significant this seems like a reasonable trade-off:

```
Baseline, -Xint
Benchmark                  Mode  Cnt     Score     Error  Units
Wrappers.forBasicType      avgt    5  1246.144 ? 149.933  ns/op
Wrappers.forPrimitive      avgt    5  4955.297 ? 329.869  ns/op
Wrappers.forPrimitiveType  avgt    5   716.840 ?  62.568  ns/op
Wrappers.forWrapper        avgt    5  5774.700 ? 367.627  ns/op

Patch, -Xint
Benchmark                  Mode  Cnt     Score     Error  Units
Wrappers.forBasicType      avgt    5  1068.096 ? 101.728  ns/op
Wrappers.forPrimitive      avgt    5  1146.670 ?  59.142  ns/op
Wrappers.forPrimitiveType  avgt    5   998.037 ? 118.144  ns/op
Wrappers.forWrapper        avgt    5  3581.226 ?  20.167  ns/op
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284880](https://bugs.openjdk.java.net/browse/JDK-8284880): Re-examine sun.invoke.util.Wrapper hash tables

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to d21330bbce82cbfec17f5f11a7bb1f90191316c5
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 97eec9d3872b00c060f3b2e5854bdb42bd96e43b

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8242/head:pull/8242` \
`$ git checkout pull/8242`

Update a local copy of the PR: \
`$ git checkout pull/8242` \
`$ git pull https://git.openjdk.java.net/jdk pull/8242/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8242`

View PR using the GUI difftool: \
`$ git pr show -t 8242`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8242.diff">https://git.openjdk.java.net/jdk/pull/8242.diff</a>

</details>

### CODE REVIEW

File: make/test/BuildMicrobenchmark.gmk [hunk 1]
```diff
@@ -94,7 +94,7 @@ $(eval $(call SetupJavaCompilation, BUILD_JDK_MICROBENCHMARK, \
     DISABLED_WARNINGS := processing rawtypes cast serial, \
     SRC := $(MICROBENCHMARK_SRC), \
     BIN := $(MICROBENCHMARK_CLASSES), \
-    JAVAC_FLAGS := --add-exports java.base/sun.security.util=ALL-UNNAMED, \
+    JAVAC_FLAGS := --add-exports java.base/sun.security.util=ALL-UNNAMED --add-exports java.base/sun.invoke.util=ALL-UNNAMED, \
```
Comment (erikj79) [inline make/test/BuildMicrobenchmark.gmk:97]:
```suggestion
    JAVAC_FLAGS := --add-exports java.base/sun.security.util=ALL-UNNAMED \
        --add-exports java.base/sun.invoke.util=ALL-UNNAMED, \
```

**Review by erikj79 — APPROVED** [review]
Build change looks ok.

/reviewers 2 reviewer

**Review by mlchung — APPROVED** [review]
Looks good.    The copyright header end-year needs update before you push.

**Comment by cl4es** [comment]
Thanks for reviews!
/integrate

**Pull request closed without merging**